### PR TITLE
show-build-info (lib:Cabal part)

### PR DESCRIPTION
**Pull request by hvr**
This is the lib:Cabal part extracted from PR #5954 by @fendor and squashed into two commits.

---
Please include the following checklist in your PR:

* [ ] Patches conform to the [coding conventions](https://github.com/haskell/cabal/blob/master/CONTRIBUTING.md#conventions).
* [ ] Any changes that could be relevant to users have been recorded in the changelog.
* [ ] The documentation has been updated, if necessary.

Please also shortly describe how you tested your change. Bonus points for added tests!

### CODE REVIEW

File: Cabal/Distribution/Simple/ShowBuildInfo.hs [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+-- linking against Cabal. This is an important advantage as direct API usage
+-- tends to be rather fragile in the presence of user-initiated upgrades of
+-- Cabal.
+--
+-- Below is an example of the output this module produces,
+--
+-- @
+-- { "cabal-version": "1.23.0.0",
+--   "compiler": {
+--     "flavor": "GHC",
```
Comment (phadej) [inline Cabal/Distribution/Simple/ShowBuildInfo.hs]: What would tea-drinking-dragon say seeing this one!

File: Cabal/Distribution/Simple/ShowBuildInfo.hs [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+        flavorToProgram GHC = Just ghcProgram
+        flavorToProgram GHCJS = Just ghcjsProgram
+        flavorToProgram UHC = Just uhcProgram
+        flavorToProgram JHC = Just jhcProgram
+        flavorToProgram _ = Nothing
+
+    mkComponentInfo (name, clbi) = JsonObject
+      [ "type" .= JsonString compType
+      , "name" .= JsonString (prettyShow name)
+      , "unit-id" .= JsonString (prettyShow $ componentUnitId clbi)
```
Comment (phadej) [inline Cabal/Distribution/Simple/ShowBuildInfo.hs]: this is not mentioned in the module header docs, but this is an important bit!

File: Cabal/Distribution/Simple/Utils/Json.hs [hunk 3]
```diff
@@ -0,0 +1,46 @@
+-- | Utility json lib for Cabal
+-- TODO: Remove it again.
```
Comment (phadej) [inline Cabal/Distribution/Simple/Utils/Json.hs]: please explain. Make an issue, link to it.

TODOs in the code are almost never fixed "immediately".

File: Cabal/Distribution/Simple/ShowBuildInfo.hs (hunk 1 above)
Comment (23Skidoo) [reply Cabal/Distribution/Simple/ShowBuildInfo.hs]: The `lib:Cabal` data type is called `CompilerFlavor`... in principle we can change the name for 3.0.

**Comment by 23Skidoo** [comment]
OK, let's merge now, but please expand comments as requested in the review feedback.

**Comment by 23Skidoo** [comment]
Merged, thanks!

File: Cabal/Distribution/Simple/ShowBuildInfo.hs [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+                -> ComponentLocalBuildInfo
+                -> [String]
+getCompilerArgs bi lbi clbi =
+  case compilerFlavor $ compiler lbi of
+      GHC   -> ghc
+      GHCJS -> ghc
+      c     -> error $ "ShowBuildInfo.getCompilerArgs: Don't know how to get "++
+                       "build arguments for compiler "++show c
+  where
+    -- This is absolutely awful
```
[the remark this answers is not in the document]
Comment (andreabedini) [reply Cabal/Distribution/Simple/ShowBuildInfo.hs:155]: I wish there was.

**Pull request merged**